### PR TITLE
Unitary system now reports warning when cycling supply fan schedule is used with setpoint control type

### LINKED ISSUE #8285: Near zero single speed cooling coil electric power for unitary system with cycling supply fan
Issue overview
--------------
I have a model that has a Unitary System set up as a DOAS with gas heating, DX cooling, and cycling fan operation. The cycling fan operation is set by the Supply Air Fan Operating Mode Schedule Name input field of the unitary system object with the schedule set to always zero. 

image
image

According to IO Ref for this input field:
Supply Air Fan Operating Mode Schedule Name - This alpha field specifies the name of the supply air fan operating mode schedule. The supply air fan operating mode may vary during the simulation based on time-of-day or with a change of season. Schedule values of 0 denote that the unitary system supply air fan and the heating or cooling coil cycle on and off together to meet the heating or cooling load (a.k.a. AUTO fan). Schedule values other than 0 denote that the supply fan runs continuously while the heating or cooling coil cycles to meet the load. The SingleZoneVAV control type is only active when the supply air fan runs continuously (i.e., during cycling fan operation the Control Type = Load model is used).

Simulating just a heating and cooling design day, the DOAS Unitary system works correctly on both days. However, in the cooling design day the Cooling Coil Electricity Rate is near zero. See table below.

image

If the fan operation schedule changes to always 1, indicating a constant flow supply fan per IO Ref, then the Cooling Coil Electricity Rate is much more realistic.

[CyclingFan_SetpointControl.zip](https://github.com/NREL/EnergyPlus/files/5453260/CyclingFan_SetpointControl.zip)

### Details
Some additional details for this issue (if relevant):
 - EnergyPlus v9.3

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by matthew-larson**
Pull request overview
---------------------
 - Fixes #8285 
 - This pull request corrects a check for the unitary system based on the Supply Air Fan Operating Mode Schedule when the Control Type is set to Setpoint.  Previously, the fan operation mode schedule was not being checked, which was keeping the cooling coil from being enabled.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: Well, this works but now this same code is in 2 places, here and in initLoadBasedControl. Shouldn't these get moved to initUnitarySystems inside the FirstHVACIteration test (line 1084) so that both control types execute this, it's only done once per time step and there is only 1 copy of these lines of code.

Comment (matthew-larson) [reply src/EnergyPlus/UnitarySystem.cc]: That makes sense, I'll make this adjustment once the CI finishes.  Thanks

**Comment by matthew-larson** [comment]
@rraustad I moved the fix to initUnitarySystems as recommended to eliminate the duplicated check.  If you could make sure I got it in the right spot, it's ready for another review.

File: tst/EnergyPlus/unit/UnitarySystem.unit.cc
Comment (rraustad) [inline tst/EnergyPlus/unit/UnitarySystem.unit.cc]: Shoot, this was ready to merge. Line feed needed at end of unit test.

**Review by rraustad — CHANGES_REQUESTED** [review]
Revision to code location so load based and set point based control use same logic. This looks good. Unit test still needs mod to last line.

File: src/EnergyPlus/UnitarySystem.cc
Comment (rraustad) [inline src/EnergyPlus/UnitarySystem.cc]: Well, I actually suggested this be moved to line 1006, or better yet 1010, below since it needs only be done once per time step. Might as well make that change while you correct the unit test.

**Comment by rraustad** [comment]
Perfect, I think I'll run the defect file before merging this. Also, check off the pull request author items in the pull request description. Think about the PR title from the users perspective.

**Comment by rraustad** [comment]
Thanks @matthew-larson. You are done with this one unless something else unexpectedly pops up. I'll get to testing the defect file when I can.

**Comment by matthew-larson** [comment]
Sounds good, thanks for your help and walking me through the code, it was very valuable.  I'll keep tabs on this to make sure everything comes back clean.

**Comment by matthew-larson** [comment]
@rraustad I updated the pull request for the new fix of making sure the correct error is caught when a cycling fan is used with Setpoint control.  I also updated the unit test to check for the error.  Let me know if you have any additional comments on this.

File: tst/EnergyPlus/unit/UnitarySystem.unit.cc
Comment (rraustad) [inline tst/EnergyPlus/unit/UnitarySystem.unit.cc:15561]: @matthew-larson are you clang formatting as you go? It seems this line would have wrapped. Not sure if it's worth another push if everything else is OK.

Comment (matthew-larson) [reply tst/EnergyPlus/unit/UnitarySystem.unit.cc:15561]: I have clang formatting ON in CLion.  I can double check that it's working right.

**Comment by rraustad** [comment]
This seems correct now.

The DOAS unit uses setpoint control and FanOpMode schedule value = 0

    ** Severe  ** AirLoopHVAC:UnitarySystem = DOAS UNITARY
    **   ~~~   ** For FAN:ONOFF = DOAS UNITARY SUPPLY FAN
    **   ~~~   ** Fan operating mode must be continuous (fan operating mode schedule values > 0).
    **   ~~~   ** Error found in Supply Air Fan Operating Mode Schedule Name DOAS UNITARY FAN MODE SCHEDULE
    **   ~~~   ** ...schedule values must be (>0., <=1.)
    **  Fatal  ** getUnitarySystemInputData: previous errors cause termination. Check inputs

If value for FanOpMode schedule is changed to 1 or that input is left blank then there are no fan op mode schedule warnings and the coil turns on:

image

The check of the FanOpMode schedule for both set point control and load or SZVAV control is not necessary so I'm glad that code was not ultimately moved as discussed. For set point control, we know the fan op mode must be constant so only checking once that the schedule value is > 0 is the only thing needed.

**Comment by rraustad** [comment]
Running unit tests locally multiple times to get a clean pass. Keeps stalling at LoggingSubStep4stepsPerHour. That unit test should not be related to this change.

**Comment by mitchute** [comment]
> Keeps stalling at LoggingSubStep4stepsPerHour

@rraustad, FWIW, that test usually takes a few minutes to complete on my machine.

**Comment by rraustad** [comment]
Yeah, that unit test finally finished after 15 minutes, using a release build.

image

**Review by rraustad — APPROVED** [review]
Simple changes made to catch issues with fan operating mode schedule when set point control is used.

**Comment by mitchute** [comment]
15 minutes... 😮

**Pull request merged**
